haproxy-ingress canary dumps core when trying to use ssl-passthrough

When creating an ingress with the `ssl-passthrough` annotation, the haproxy pod crashes with this log output:
```
I1130 22:45:59.362355       5 launch.go:142] 
Name:       HAProxy
Release:    v0.7-beta.2
Build:      git-8875595
Repository: https://github.com/jcmoraisjr/haproxy-ingress
I1130 22:45:59.362404       5 launch.go:145] Watching for ingress class: haproxy
I1130 22:45:59.362562       5 launch.go:345] Creating API client for https://10.96.0.1:443
I1130 22:45:59.367565       5 launch.go:357] Running in Kubernetes Cluster version v1.10 (v1.10.0) - git (clean) commit fc32d2f3698e36b93322a3465f63a14e9f0eaead - platform linux/amd64
I1130 22:45:59.368931       5 launch.go:169] validated default/ingress-default-backend as the default backend
I1130 22:45:59.373212       5 controller.go:1483] starting Ingress controller
I1130 22:45:59.380613       5 event.go:218] Event(v1.ObjectReference{Kind:"Ingress", Namespace:"default", Name:"vault", UID:"5d353da1-f4f1-11e8-b72e-0800277849aa", APIVersion:"extensions", ResourceVersion:"139744", FieldPath:""}): type: 'Normal' reason: 'CREATE' Ingress default/vault
I1130 22:46:04.674928       5 controller.go:1491] running initial sync of secrets
W1130 22:46:04.677344       5 controller.go:893] ignoring path '%v' from sslpassthrough ingress %v/%vdefaultvault
I1130 22:46:04.678779       5 leaderelection.go:174] attempting to acquire leader lease...
E1130 22:46:04.679447       5 runtime.go:66] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/home/travis/gopath/src/github.com/jcmoraisjr/haproxy-ingress/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:72
/home/travis/gopath/src/github.com/jcmoraisjr/haproxy-ingress/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/home/travis/gopath/src/github.com/jcmoraisjr/haproxy-ingress/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/home/travis/.gimme/versions/go1.9.2.linux.amd64/src/runtime/asm_amd64.s:509
/home/travis/.gimme/versions/go1.9.2.linux.amd64/src/runtime/panic.go:491
/home/travis/.gimme/versions/go1.9.2.linux.amd64/src/runtime/panic.go:63
/home/travis/.gimme/versions/go1.9.2.linux.amd64/src/runtime/signal_unix.go:367
/home/travis/gopath/src/github.com/jcmoraisjr/haproxy-ingress/pkg/common/ingress/controller/controller.go:636
/home/travis/gopath/src/github.com/jcmoraisjr/haproxy-ingress/pkg/common/ingress/controller/controller.go:295
/home/travis/gopath/src/github.com/jcmoraisjr/haproxy-ingress/pkg/common/ingress/controller/controller.go:177
/home/travis/gopath/src/github.com/jcmoraisjr/haproxy-ingress/pkg/common/task/queue.go:112
/home/travis/gopath/src/github.com/jcmoraisjr/haproxy-ingress/pkg/common/task/queue.go:59
/home/travis/gopath/src/github.com/jcmoraisjr/haproxy-ingress/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133
/home/travis/gopath/src/github.com/jcmoraisjr/haproxy-ingress/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134
/home/travis/gopath/src/github.com/jcmoraisjr/haproxy-ingress/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
/home/travis/gopath/src/github.com/jcmoraisjr/haproxy-ingress/pkg/common/task/queue.go:59
/home/travis/.gimme/versions/go1.9.2.linux.amd64/src/runtime/asm_amd64.s:2337
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xfbc8a8]

goroutine 64 [running]:
github.com/jcmoraisjr/haproxy-ingress/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/home/travis/gopath/src/github.com/jcmoraisjr/haproxy-ingress/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x111
panic(0x1126b60, 0x1b2bec0)
	/home/travis/.gimme/versions/go1.9.2.linux.amd64/src/runtime/panic.go:491 +0x283
github.com/jcmoraisjr/haproxy-ingress/pkg/common/ingress/controller.(*GenericController).getBackendServers(0xc420358aa0, 0xc4202b8000, 0x1, 0x1, 0x1, 0xc4202b8000, 0x0, 0x1, 0x7ff7b6954b50, 0x6)
	/home/travis/gopath/src/github.com/jcmoraisjr/haproxy-ingress/pkg/common/ingress/controller/controller.go:636 +0x2e8
github.com/jcmoraisjr/haproxy-ingress/pkg/common/ingress/controller.(*GenericController).syncIngress(0xc420358aa0, 0x116bca0, 0xc4202c8080, 0xc428410402, 0x13ff02058)
	/home/travis/gopath/src/github.com/jcmoraisjr/haproxy-ingress/pkg/common/ingress/controller/controller.go:295 +0x38d
github.com/jcmoraisjr/haproxy-ingress/pkg/common/ingress/controller.(*GenericController).(github.com/jcmoraisjr/haproxy-ingress/pkg/common/ingress/controller.syncIngress)-fm(0x116bca0, 0xc4202c8080, 0xa, 0xc420500d68)
	/home/travis/gopath/src/github.com/jcmoraisjr/haproxy-ingress/pkg/common/ingress/controller/controller.go:177 +0x3e
github.com/jcmoraisjr/haproxy-ingress/pkg/common/task.(*Queue).worker(0xc420236b40)
	/home/travis/gopath/src/github.com/jcmoraisjr/haproxy-ingress/pkg/common/task/queue.go:112 +0x35b
github.com/jcmoraisjr/haproxy-ingress/pkg/common/task.(*Queue).(github.com/jcmoraisjr/haproxy-ingress/pkg/common/task.worker)-fm()
	/home/travis/gopath/src/github.com/jcmoraisjr/haproxy-ingress/pkg/common/task/queue.go:59 +0x2a
github.com/jcmoraisjr/haproxy-ingress/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc420029fa8)
	/home/travis/gopath/src/github.com/jcmoraisjr/haproxy-ingress/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x5e
github.com/jcmoraisjr/haproxy-ingress/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc4204a9fa8, 0x3b9aca00, 0x0, 0x1, 0xc42022e8a0)
	/home/travis/gopath/src/github.com/jcmoraisjr/haproxy-ingress/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0xbd
github.com/jcmoraisjr/haproxy-ingress/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc420029fa8, 0x3b9aca00, 0xc42022e8a0)
	/home/travis/gopath/src/github.com/jcmoraisjr/haproxy-ingress/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
github.com/jcmoraisjr/haproxy-ingress/pkg/common/task.(*Queue).Run(0xc420236b40, 0x3b9aca00, 0xc42022e8a0)
	/home/travis/gopath/src/github.com/jcmoraisjr/haproxy-ingress/pkg/common/task/queue.go:59 +0x55
created by github.com/jcmoraisjr/haproxy-ingress/pkg/common/ingress/controller.(*GenericController).Start
	/home/travis/gopath/src/github.com/jcmoraisjr/haproxy-ingress/pkg/common/ingress/controller/controller.go:1504 +0x3df
```

--- a/pkg/common/ingress/controller/controller.go
+++ b/pkg/common/ingress/controller/controller.go
@@ -890,7 +890,7 @@ func (ic *GenericController) createUpstreams(data []*extensions.Ingress, du *ing
 				backends := []extensions.IngressBackend{path.Backend}
 				if sslpt.HasSSLPassthrough {
 					if path.Path != "" && path.Path != rootLocation {
-						glog.Warning(
+						glog.Warningf(
 							"ignoring path '%v' from sslpassthrough ingress %v/%v",
 							path.Path, ing.Namespace, ing.Name)
 						continue

--- a/pkg/common/ingress/controller/controller.go
+++ b/pkg/common/ingress/controller/controller.go
@@ -608,7 +608,11 @@ func (ic *GenericController) getBackendServers(ingresses []*extensions.Ingress) 
 					path.Backend.ServiceName,
 					path.Backend.ServicePort.String())
 
-				ups := upstreams[upsName]
+				ups, found := upstreams[upsName]
+				if !found {
+					// a skipped ssl-passthrough path, just ignore
+					continue
+				}
 
 				var upshttpPassName string
 				if server.SSLPassthrough.HasSSLPassthrough && server.SSLPassthrough.HTTPPort > 0 {

--- a/pkg/common/ingress/controller/controller.go
+++ b/pkg/common/ingress/controller/controller.go
@@ -301,7 +301,7 @@ func (ic *GenericController) syncIngress(item interface{}) error {
 		}
 
 		for _, loc := range server.Locations {
-			if loc.Path != rootLocation {
+			if loc.Path != "" && loc.Path != rootLocation {
 				glog.Warningf("ignoring path %v of ssl passthrough host %v", loc.Path, server.Hostname)
 				continue
 			}
@@ -889,7 +889,7 @@ func (ic *GenericController) createUpstreams(data []*extensions.Ingress, du *ing
 			for _, path := range rule.HTTP.Paths {
 				backends := []extensions.IngressBackend{path.Backend}
 				if sslpt.HasSSLPassthrough {
-					if path.Path != rootLocation {
+					if path.Path != "" && path.Path != rootLocation {
 						glog.Warning(
 							"ignoring path '%v' from sslpassthrough ingress %v/%v",
 							path.Path, ing.Namespace, ing.Name)

